In [91]:
import requests as rq
import pandas as pd
import linecache 



## Gathering ##

As requested in the project details, I get the data from the following sources:

### csv source: WeRateDogs Twitter archive  ### 

In [92]:
archive_df = pd.read_csv('twitter-archive-enhanced.csv')

### tsv source: image predictions ###

In [93]:
img_predictions = 'https://d17h27t6h515a5.cloudfront.net/topher/2017/August/599fd2ad_image-predictions/image-predictions.tsv'
r = rq.get(img_predictions)

In [94]:
r.status_code

200

In [95]:
with open('image-predictions.tsv', 'wb') as f:
        f.write(r.content)

In [96]:
img_predictions_df  = pd.read_csv('image-predictions.tsv', sep= '\t')

### json source: querying the Twitter API ###

Using the code facilitated for this project (`twitter_api.py`), I downloaded json data for the tweet IDs contained in source one above. Once in the text file, I used the read_json function of Pandas:

In [97]:
json_df = pd.read_json('tweet_json.txt', lines = True )

## Assesing ##
> Detect and document at least eight (8) quality issues 
 and two (2) tidiness issues in your wrangle_act.ipynb Jupyter Notebook.
 To meet specifications, the issues that satisfy the Project Motivation (see the Key Points header on the previous page) must be assessed.

First, I proceed to look at the essential information concerning the three dataframes I just created. I do this, by invoking `info()` and `shape` for each df.

In [98]:
archive_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2356 entries, 0 to 2355
Data columns (total 17 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   tweet_id                    2356 non-null   int64  
 1   in_reply_to_status_id       78 non-null     float64
 2   in_reply_to_user_id         78 non-null     float64
 3   timestamp                   2356 non-null   object 
 4   source                      2356 non-null   object 
 5   text                        2356 non-null   object 
 6   retweeted_status_id         181 non-null    float64
 7   retweeted_status_user_id    181 non-null    float64
 8   retweeted_status_timestamp  181 non-null    object 
 9   expanded_urls               2297 non-null   object 
 10  rating_numerator            2356 non-null   int64  
 11  rating_denominator          2356 non-null   int64  
 12  name                        2356 non-null   object 
 13  doggo                       2356 

In [129]:
archive_df.text.head()

0    This is Phineas. He's a mystical boy. Only eve...
1    This is Tilly. She's just checking pup on you....
2    This is Archie. He is a rare Norwegian Pouncin...
3    This is Darla. She commenced a snooze mid meal...
4    This is Franklin. He would like you to stop ca...
Name: text, dtype: object

In [99]:
archive_df.shape

(2356, 17)

In [100]:
img_predictions_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2075 entries, 0 to 2074
Data columns (total 12 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   tweet_id  2075 non-null   int64  
 1   jpg_url   2075 non-null   object 
 2   img_num   2075 non-null   int64  
 3   p1        2075 non-null   object 
 4   p1_conf   2075 non-null   float64
 5   p1_dog    2075 non-null   bool   
 6   p2        2075 non-null   object 
 7   p2_conf   2075 non-null   float64
 8   p2_dog    2075 non-null   bool   
 9   p3        2075 non-null   object 
 10  p3_conf   2075 non-null   float64
 11  p3_dog    2075 non-null   bool   
dtypes: bool(3), float64(3), int64(2), object(4)
memory usage: 152.1+ KB


In [101]:
img_predictions_df.shape

(2075, 12)

We see all entries in the predicitions data frame has a valid URL 
#TODO find a proper URL regex and pass it to count()

In [102]:
img_predictions_df.jpg_url.str.count('http').size

2075

In [103]:
json_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2331 entries, 0 to 2330
Data columns (total 32 columns):
 #   Column                         Non-Null Count  Dtype              
---  ------                         --------------  -----              
 0   created_at                     2331 non-null   datetime64[ns, UTC]
 1   id                             2331 non-null   int64              
 2   id_str                         2331 non-null   int64              
 3   full_text                      2331 non-null   object             
 4   truncated                      2331 non-null   bool               
 5   display_text_range             2331 non-null   object             
 6   entities                       2331 non-null   object             
 7   extended_entities              2059 non-null   object             
 8   source                         2331 non-null   object             
 9   in_reply_to_status_id          77 non-null     float64            
 10  in_reply_to_status_id_st

In [104]:
json_df.shape

(2331, 32)

The information above leaves us that the image predictions data frame has the least information with 2075 rows, next we have the json one with 2351, and finally the archive with 2356.

### Visual assessment ####
With our dataframes loaded, we can do a quick visual assessment, to get familiar with the data:

In [122]:
archive_df.sample(5)

tweet_id  in_reply_to_status_id  in_reply_to_user_id  \
1423  697995514407682048                    NaN                  NaN   
1783  677673981332312066                    NaN                  NaN   
1905  674606911342424069           6.744689e+17         4.196984e+09   
2247  667873844930215936                    NaN                  NaN   
37    885167619883638784                    NaN                  NaN   

                      timestamp  \
1423  2016-02-12 04:07:53 +0000   
1783  2015-12-18 02:17:22 +0000   
1905  2015-12-09 15:09:55 +0000   
2247  2015-11-21 01:15:07 +0000   
37    2017-07-12 16:03:00 +0000   

                                                 source  \
1423  <a href="http://twitter.com/download/iphone" r...   
1783  <a href="http://twitter.com/download/iphone" r...   
1905  <a href="http://twitter.com/download/iphone" r...   
2247  <a href="http://twitter.com/download/iphone" r...   
37    <a href="http://twitter.com/download/iphone" r...   

                                                   text  retweeted_status_id  \
1423  "Dammit hooman quit playin I jus wanna wheat t...                  NaN   
1783  Endangered triangular pup here. Could be a wiz...                  NaN   
1905  The 13/10 also takes into account this impecca...                  NaN   
2247  Neat dog. Lots of spikes. Always in push-up po...                  NaN   
37    Here we have a corgi undercover as a malamute....                  NaN   

      retweeted_status_user_id retweeted_status_timestamp  \
1423                       NaN                        NaN   
1783                       NaN                        NaN   
1905                       NaN                        NaN   
2247                       NaN                        NaN   
37                         NaN                        NaN   

                                          expanded_urls  rating_numerator  \
1423  https://twitter.com/dog_rates/status/697995514...                11   
1783  https://twitter.com/dog_rates/status/677673981...                 9   
1905                                                NaN                13   
2247  https://twitter.com/dog_rates/status/667873844...                10   
37    https://twitter.com/dog_rates/status/885167619...                13   

      rating_denominator  name doggo floofer pupper puppo  
1423                  10  None  None    None   None  None  
1783                  10  None  None    None   None  None  
1905                  10  None  None    None   None  None  
2247                  10  None  None    None   None  None  
37                    10  None  None    None   None  None

In [128]:
display(archive_df.rating_numerator.value_counts(ascending=False))
display(archive_df.rating_denominator.value_counts(ascending=False))

12      558
11      464
10      461
13      351
9       158
8       102
7        55
14       54
5        37
6        32
3        19
4        17
1         9
2         9
0         2
15        2
75        2
420       2
182       1
204       1
143       1
121       1
99        1
20        1
45        1
27        1
17        1
24        1
26        1
44        1
50        1
60        1
80        1
84        1
88        1
1776      1
960       1
666       1
144       1
165       1
Name: rating_numerator, dtype: int64

10     2333
11        3
50        3
20        2
80        2
0         1
120       1
7         1
170       1
150       1
130       1
90        1
110       1
2         1
70        1
40        1
16        1
15        1
Name: rating_denominator, dtype: int64

In [106]:
display(archive_df.doggo.value_counts(),  archive_df.floofer.value_counts())

None     2259
doggo      97
Name: doggo, dtype: int64

None       2346
floofer      10
Name: floofer, dtype: int64

In [107]:
display(archive_df.pupper.value_counts(), archive_df.puppo.value_counts())

None      2099
pupper     257
Name: pupper, dtype: int64

None     2326
puppo      30
Name: puppo, dtype: int64

In [108]:
predicate = archive_df.retweeted_status_id.isnull() == False
archive_df[predicate][['retweeted_status_id','retweeted_status_user_id','timestamp','retweeted_status_timestamp']]

retweeted_status_id  retweeted_status_user_id  \
19           8.874740e+17              4.196984e+09   
32           8.860537e+17              1.960740e+07   
36           8.305833e+17              4.196984e+09   
68           8.780576e+17              4.196984e+09   
73           8.782815e+17              4.196984e+09   
...                   ...                       ...   
1023         6.678667e+17              4.196984e+09   
1043         6.671383e+17              4.196984e+09   
1242         7.119983e+17              7.832140e+05   
2259         6.675487e+17              4.296832e+09   
2260         6.675484e+17              4.296832e+09   

                      timestamp retweeted_status_timestamp  
19    2017-07-21 01:02:36 +0000  2017-07-19 00:47:34 +0000  
32    2017-07-15 02:45:48 +0000  2017-07-15 02:44:07 +0000  
36    2017-07-13 01:35:06 +0000  2017-02-12 01:04:29 +0000  
68    2017-06-26 00:13:58 +0000  2017-06-23 01:10:23 +0000  
73    2017-06-24 00:09:53 +0000  2017-06-23 16:00:04 +0000  
...                         ...                        ...  
1023  2016-06-25 01:52:36 +0000  2015-11-21 00:46:50 +0000  
1043  2016-06-17 16:01:16 +0000  2015-11-19 00:32:12 +0000  
1242  2016-03-21 19:31:59 +0000  2016-03-21 19:29:52 +0000  
2259  2015-11-20 03:51:52 +0000  2015-11-20 03:43:06 +0000  
2260  2015-11-20 03:51:47 +0000  2015-11-20 03:41:59 +0000  

[181 rows x 4 columns]

In [125]:
img_predictions_df.sample(5)

tweet_id                                          jpg_url  \
440   674447403907457024  https://pbs.twimg.com/media/CVweVUfW4AACPwI.jpg   
2022  881268444196462592  https://pbs.twimg.com/media/DDrk-f9WAAI-WQv.jpg   
131   668297328638447616  https://pbs.twimg.com/media/CUZE4IWW4AAZmDf.jpg   
920   701952816642965504  https://pbs.twimg.com/media/Cb3WXMUUMAIuzL8.jpg   
1332  757725642876129280  https://pbs.twimg.com/media/CoP7c4bWcAAr55g.jpg   

      img_num                 p1   p1_conf  p1_dog                  p2  \
440         1  Brabancon_griffon  0.409909    True            malinois   
2022        1             tusker  0.473303   False     Indian_elephant   
131         1       king_penguin  0.606747   False            ice_bear   
920         1         toy_poodle  0.331707    True    miniature_poodle   
1332        2          seat_belt  0.425176   False  Labrador_retriever   

       p2_conf  p2_dog               p3   p3_conf  p3_dog  
440   0.244649    True     bull_mastiff  0.074819    True  
2022  0.245646   False             ibex  0.055661   False  
131   0.264221   False       Eskimo_dog  0.032784    True  
920   0.272485    True  standard_poodle  0.169415    True  
1332  0.128128    True      Siamese_cat  0.091241   False

In [126]:
img_predictions_df[img_predictions_df['p1_dog'] == False].sample(5)

tweet_id                                          jpg_url  \
1038  712065007010385924  https://pbs.twimg.com/media/CeHDV73W0AM5Cf8.jpg   
95    667550904950915073  https://pbs.twimg.com/media/CUOb_gUUkAACXdS.jpg   
1206  742150209887731712  https://pbs.twimg.com/media/CkylrVWWsAAiXJE.jpg   
1891  849051919805034497  https://pbs.twimg.com/media/C8hwNxbXYAAwyVG.jpg   
1638  806629075125202948  https://pbs.twimg.com/media/CzG425nWgAAnP7P.jpg   

      img_num             p1   p1_conf  p1_dog                   p2   p2_conf  \
1038        1          goose  0.214301   False               gibbon  0.084253   
95          1       web_site  0.999335   False               vizsla  0.000081   
1206        1    Siamese_cat  0.112413   False       French_bulldog  0.071414   
1891        1       fountain  0.997509   False  American_black_bear  0.001413   
1638        2  Arabian_camel  0.366248   False          house_finch  0.209852   

      p2_dog              p3   p3_conf  p3_dog  
1038   False           pizza  0.080168   False  
95      True          collie  0.000069    True  
1206    True             hog  0.062465   False  
1891   False         sundial  0.000681   False  
1638   False  cocker_spaniel  0.046403    True

In [110]:
img_predictions_df.img_num.value_counts()

1    1780
2     198
3      66
4      31
Name: img_num, dtype: int64

In [111]:
json_df.sample(5)

created_at                  id              id_str  \
243  2017-03-22 16:04:20+00:00  844580511645339650  844580511645339648   
1145 2016-04-15 01:26:47+00:00  720785406564900865  720785406564900864   
2073 2015-11-28 22:15:21+00:00  670727704916926465  670727704916926464   
1734 2015-12-21 00:53:29+00:00  678740035362037760  678740035362037760   
2004 2015-12-02 00:58:41+00:00  671855973984772097  671855973984772096   

                                              full_text  truncated  \
243   This is Charlie. He wants to know if you have ...      False   
1145  This is Archie. He hears everything you say. D...      False   
2073  This is Jeffrie. He's a handheld pup. Excellen...      False   
1734  Meet Tango. He's a large dog. Doesn't care muc...      False   
2004  Remarkable dog here. Walks on back legs really...      False   

     display_text_range                                           entities  \
243            [0, 129]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1145           [0, 104]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2073           [0, 121]  {'hashtags': [], 'symbols': [], 'user_mentions...   
1734           [0, 137]  {'hashtags': [], 'symbols': [], 'user_mentions...   
2004           [0, 117]  {'hashtags': [], 'symbols': [], 'user_mentions...   

                                      extended_entities  \
243   {'media': [{'id': 844580505345511424, 'id_str'...   
1145  {'media': [{'id': 720785400575455232, 'id_str'...   
2073  {'media': [{'id': 670727697677606912, 'id_str'...   
1734  {'media': [{'id': 678740029439614976, 'id_str'...   
2004  {'media': [{'id': 671855970818121728, 'id_str'...   

                                                 source  \
243   <a href="http://twitter.com/download/iphone" r...   
1145  <a href="http://twitter.com/download/iphone" r...   
2073  <a href="http://twitter.com/download/iphone" r...   
1734  <a href="http://twitter.com/download/iphone" r...   
2004  <a href="http://twitter.com/download/iphone" r...   

      in_reply_to_status_id  ...  favorited  retweeted  possibly_sensitive  \
243                     NaN  ...      False      False                 0.0   
1145                    NaN  ...      False      False                 0.0   
2073                    NaN  ...      False      False                 0.0   
1734                    NaN  ...      False      False                 0.0   
2004                    NaN  ...      False      False                 0.0   

     possibly_sensitive_appealable lang  retweeted_status  quoted_status_id  \
243                            0.0   en               NaN               NaN   
1145                           0.0   en               NaN               NaN   
2073                           0.0   en               NaN               NaN   
1734                           0.0   en               NaN               NaN   
2004                           0.0   en               NaN               NaN   

     quoted_status_id_str  quoted_status_permalink  quoted_status  
243                   NaN                      NaN            NaN  
1145                  NaN                      NaN            NaN  
2073                  NaN                      NaN            NaN  
1734                  NaN                      NaN            NaN  
2004                  NaN                      NaN            NaN  

[5 rows x 32 columns]

In [112]:
#json_df[json_df.retweeted_status_id.isnull() == False]#.sample(3)
#json_df[.value_counts() ==]
display('favorite count stats')
display( json_df.favorite_count.describe())
display('retweet count stats')
display(json_df.retweet_count.describe())

'favorite count stats'

count      2331.000000
mean       7319.238953
std       11367.525636
min           0.000000
25%        1269.500000
50%        3174.000000
75%        8947.000000
max      150692.000000
Name: favorite_count, dtype: float64

'retweet count stats'

count     2331.000000
mean      2591.102531
std       4383.211411
min          1.000000
25%        525.500000
50%       1209.000000
75%       3000.000000
max      74376.000000
Name: retweet_count, dtype: float64

In [113]:
json_df.retweeted.value_counts()

False    2331
Name: retweeted, dtype: int64

In [117]:
json_df[['created_at', 'id']].head()

created_at                  id
0 2017-08-01 16:23:56+00:00  892420643555336193
1 2017-08-01 00:17:27+00:00  892177421306343426
2 2017-07-31 00:18:03+00:00  891815181378084864
3 2017-07-30 15:58:51+00:00  891689557279858688
4 2017-07-29 16:00:24+00:00  891327558926688256

In [119]:
archive_df[['timestamp', 'tweet_id']].head()

timestamp            tweet_id
0  2017-08-01 16:23:56 +0000  892420643555336193
1  2017-08-01 00:17:27 +0000  892177421306343426
2  2017-07-31 00:18:03 +0000  891815181378084864
3  2017-07-30 15:58:51 +0000  891689557279858688
4  2017-07-29 16:00:24 +0000  891327558926688256

In [130]:
json_df.full_text.head()

0    This is Phineas. He's a mystical boy. Only eve...
1    This is Tilly. She's just checking pup on you....
2    This is Archie. He is a rare Norwegian Pouncin...
3    This is Darla. She commenced a snooze mid meal...
4    This is Franklin. He would like you to stop ca...
Name: full_text, dtype: object

## Cleaning ##

Once identified the issues with the sources above, I start off this cleaning section, doing a backup of the "original" dataframes:

> Cleaning includes merging individual pieces of data according to the rules of tidy data.

In [12]:
archive_clean_df = archive_df
images_clean_df  = img_predictions_df
json_clean_df = json_df

## Storing ##

## Analyzing ##
>  At least three (3) insights . Some insights to obtain are:

 1) what's the dog best rated in the dataset. IN case of a tie, what;s the most recent?
 
 2) what's the dog worst rated in the dataset. IN case of a tie, what;s the most recent?
 
 3) what's the dog with more favorite tweets counts?

## Visualizing ##

In [ ]:
# At least one (1) visualization must be produced.
